agent/src/beerocks/monitor/monitor_thread.cpp

```diff
@@ -1484,6 +1484,52 @@ bool monitor_thread::hal_event_handler(bwl::base_wlan_hal::hal_event_ptr_t event
 
         // If event == Channel_Scan_New_Results_Ready do nothing since is_dump's default is 0
         if (event == Event::Channel_Scan_Dump_Result) {
+            auto msg = static_cast<bwl::sCHANNEL_SCAN_RESULTS_NOTIFICATION *>(data);
+
+            auto in_result  = &(msg->channel_scan_results);
+            auto out_result = &(notification->scan_results());
+
+            // Strings, simple arrays & vectors:
+            string_utils::copy_string(out_result->ssid, in_result->ssid,
+                                      beerocks::message::WIFI_SSID_MAX_LENGTH);
+            std::copy_n(in_result->bssid.oct, sizeof(out_result->bssid.oct), out_result->bssid.oct);
+            std::copy(in_result->basic_data_transfer_rates_kbps.begin(),
+                      in_result->basic_data_transfer_rates_kbps.end(),
+                      out_result->basic_data_transfer_rates_kbps);
+            std::copy(in_result->supported_data_transfer_rates_kbps.begin(),
+                      in_result->supported_data_transfer_rates_kbps.end(),
+                      out_result->supported_data_transfer_rates_kbps);
+
+            // Simple values
+            out_result->channel             = in_result->channel;
+            out_result->signal_strength_dBm = in_result->signal_strength_dBm;
+            out_result->beacon_period_ms    = in_result->beacon_period_ms;
+            out_result->noise_dBm           = in_result->noise_dBm;
+            out_result->dtim_period         = in_result->dtim_period;
+            out_result->channel_utilization = in_result->channel_utilization;
+
+            // Enums
+            out_result->mode = beerocks_message::eChannelScanResultMode(uint8_t(in_result->mode));
+            out_result->operating_frequency_band =
+                beerocks_message::eChannelScanResultOperatingFrequencyBand(
+                    uint8_t(in_result->operating_frequency_band));
+            out_result->operating_standards = beerocks_message::eChannelScanResultStandards(
+                uint8_t(in_result->operating_standards));
+            out_result->operating_channel_bandwidth =
+                beerocks_message::eChannelScanResultChannelBandwidth(
+                    uint8_t(in_result->operating_channel_bandwidth));
+
+            // Enum list
+            std::copy_n(reinterpret_cast<uint8_t *>(&in_result->security_mode_enabled[0]),
```

Comment:
No this can't be - instead of using `std::vector` APIs you use `reinterpret_cast`. This is wrong in so many levels.

Reply:
I could run the in_result through a std::for_each and for each value
run it through a double cast, and add it to the out_result array
example for `eChannelScanResultSecurityMode` I'd do:
```
int i = 0;
std::for_each(in_result.security_mode_enabled.begin(),
              in_result.security_mode_enabled.end(),
              [&i, &out_result](bwl::eChannelScanResultSecurityMode e) {
			      out_result.security_mode_enabled[i++] =
                      beerocks_message::eChannelScanResultSecurityMode(uint8_t(e));
              });
```

Since `in_result` holds a vector of class enums, and `out_result` holds an
array of diffrent but matching class enums, I have no other way to
convert between the two